### PR TITLE
Implement RasterLinesJoin operation

### PR DESCRIPTION
**Pull request by kellyi**
## Overview

This PR implements the RasterLinesJoin geoprocessing operation used in MapShed.

~All the infrastructure's wired up and operation's returns a count, but the count's currently slightly off.~ This is fixed now: my initial implementation was double counting values because the same row/col cells would be visited multiple times; my second try undercounted cells because it tried to throw out cells which had already been visited, but didn't track the visited cells correctly.

Now, we initialize the TrieMap to have a key which is a tuple comprising 

- a list of the raster values
- the pixel col/row
- the SpatialKey

```scala
var pixelGroups: TrieMap[(List[Int], List[Int], SpatialKey), Int] = TrieMap.empty
```

... and then we only count the cell if it hasn't already been visited in that tile

At the end we drop the cell col/row & SpatialKey...

```scala
.map { case ((key, _, _), value) => (key, value) }
```

...then group by the raster values and sum.

Connects #53

## Testing
- get this branch, then `./scripts/server` to run sbt ~reStart
- post the request json included in this zip file to `:8090/run`
[nlcd-streams-request-response.zip](https://github.com/WikiWatershed/mmw-geoprocessing/files/1260526/nlcd-streams-request-response.zip)`
- verify that the output matches the expected response
- test the RasterGroupedCount, RasterAverage, and RasterGroupedAverage operations and verify that they still return the values expected, too.

### CODE REVIEW

File: api/src/main/scala/Geoprocessing.scala
Comment (kellyi) [inline api/src/main/scala/Geoprocessing.scala]: Just occurred to me that storing the col/row as part of the TrieMap key like this could lead to undercounting if it's the case that multiple tiles would have the same `nlcdId, col, row` pixel in them. Potentially this can be avoided by storing the tile key in the key tuple, too.

Comment (kellyi) [reply api/src/main/scala/Geoprocessing.scala]: This turned out to be the problem.

**Comment by rajadain** [comment]
Taking a look now.

File: api/src/main/scala/Geoprocessing.scala
Comment (rajadain) [inline api/src/main/scala/Geoprocessing.scala]: Typo `collction` → `collection`

File: api/src/main/scala/Geoprocessing.scala
Comment (rajadain) [inline api/src/main/scala/Geoprocessing.scala]: This can be a `val`

Comment (kellyi) [inline api/src/main/scala/Geoprocessing.scala]: Probably worth replacing the second and third parts of TrieMap key -- `(List[Int], List[Int], SpatialKey)` -- here with a tuple of `(Int, Int, SpatialKey)` representing the column, row, and tile key.

File: api/src/main/scala/Geoprocessing.scala
Comment (rajadain) [inline api/src/main/scala/Geoprocessing.scala]: This `.toMap` seems extraneous

File: api/src/main/scala/Utils.scala
Comment (rajadain) [inline api/src/main/scala/Utils.scala]: Consider adding an empty line above this, as it is the value being returned.

File: api/src/main/scala/Geoprocessing.scala
Comment (rajadain) [reply api/src/main/scala/Geoprocessing.scala]: Here's a suggested implementation:

```diff
diff --git a/api/src/main/scala/Geoprocessing.scala b/api/src/main/scala/Geoprocessing.scala
index 3fb8e3e..32667a1 100644
--- a/api/src/main/scala/Geoprocessing.scala
+++ b/api/src/main/scala/Geoprocessing.scala
@@ -70,6 +70,8 @@ trait Geoprocessing extends Utils {
     ResultInt(rasterLinesJoin(rasterLayers, lines))
   }
 
+  private case class TilePixel (key: SpatialKey, col: Int, row: Int)
+
   /**
     * Given a collction of rasterLayers & a collection of lines, return the
     * values intersected by the rasterized lines.
@@ -83,7 +85,7 @@ trait Geoprocessing extends Utils {
     lines: Seq[MultiLine]
   ): Map[String, Int] = {
     val metadata = rasterLayers.head.metadata
-    var pixelGroups: TrieMap[(List[Int], List[Int], SpatialKey), Int] = TrieMap.empty
+    var pixelGroups: TrieMap[(List[Int], TilePixel), Int] = TrieMap.empty
 
     joinCollectionLayers(rasterLayers).par
       .foreach({ case (key, tiles) =>
@@ -93,8 +95,8 @@ trait Geoprocessing extends Utils {
 
         lines.par.foreach({ multiLine =>
           Rasterizer.foreachCellByMultiLineString(multiLine, re) { case (col, row) =>
-            val pixelGroup: (List[Int], List[Int], SpatialKey) =
-              (tiles.map(_.get(col, row)).toList, List(col, row), key)
+            val pixelGroup: (List[Int], TilePixel) =
+              (tiles.map(_.get(col, row)).toList, TilePixel(key, col, row))
             pixelGroups.getOrElseUpdate(pixelGroup, 1)
           }
         })
@@ -102,7 +104,7 @@ trait Geoprocessing extends Utils {
 
     pixelGroups
       .toSeq
-      .map { case ((key, _, _), value) => (key, value) }
+      .map { case ((key, _), value) => (key, value) }
       .groupBy(_._1)
       .map { case (k, v) =>
         k.toString -> v.map { case (_, n) => n }.sum
```

File: api/src/main/scala/Geoprocessing.scala
Comment (rajadain) [inline api/src/main/scala/Geoprocessing.scala]: This bit of code can be simplified to this:

```diff
diff --git a/api/src/main/scala/Geoprocessing.scala b/api/src/main/scala/Geoprocessing.scala
index 3fb8e3e..7128dce 100644
--- a/api/src/main/scala/Geoprocessing.scala
+++ b/api/src/main/scala/Geoprocessing.scala
@@ -101,13 +101,8 @@ trait Geoprocessing extends Utils {
       })
 
     pixelGroups
-      .toSeq
-      .map { case ((key, _, _), value) => (key, value) }
-      .groupBy(_._1)
-      .map { case (k, v) =>
-        k.toString -> v.map { case (_, n) => n }.sum
-      }
-      .toMap
+      .groupBy(_._1._1.toString)
+      .mapValues(_.size)
   }
 
   /**
```

File: api/src/main/scala/Geoprocessing.scala
Comment (rajadain) [reply api/src/main/scala/Geoprocessing.scala]: It may go away completely with the suggested simplification above.

**Review by rajadain — APPROVED** [review]
+1 tested. The JSON values are coming back correct. I have some code suggestions for increasing legibility. Going to test this with MMW / MapShed now!

**Comment by rajadain** [comment]
Tested in MMW against identical runs on Staging: the values are identical. Great work!

**Comment by kellyi** [comment]
Wonderful! I'm going to make the adjustments suggested above, then test everything out again before merging.

File: api/src/main/scala/Geoprocessing.scala
Comment (kellyi) [reply api/src/main/scala/Geoprocessing.scala]: Nice!

**Comment by kellyi** [comment]
Made the changes suggested above & tested again on my local. Everything's still working as before. Going to merge this once the tests pass.

Thanks for your help with this!

**Pull request merged**
